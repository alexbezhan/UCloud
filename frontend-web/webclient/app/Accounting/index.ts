import Usage from "Accounting/Usage";

export {Usage};