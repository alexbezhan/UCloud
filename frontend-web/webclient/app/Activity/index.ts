import { ActivityReduxObject } from "DefaultObjects";
import { ScrollRequest } from "Scroll";

export enum ActivityType {
    DOWNLOAD = "download",
    UPDATED = "updated",
    DELETED = "deleted",
    FAVORITE = "favorite",
    INSPECTED = "inspected",
    MOVED = "moved"
}

export interface ActivityGroup {
    type: ActivityType;
    newestTimestamp: number;
    numberOfHiddenResults: number | null;
    items: Activity[];
}

export interface Activity {
    type: ActivityType;
    timestamp: number;
    fileId: string;
    username: string;
    originalFilePath: string;
}

export interface ActivityFilter {
    collapseAt?: number;
    type?: ActivityType;
    minTimestamp?: Date;
    maxTimestamp?: Date;
}

export interface FavoriteActivity extends Activity {
    favorite: boolean;
}

export interface MovedActivity extends Activity {
    newName: string;
}


export interface ActivityDispatchProps {
    onMount: () => void;
    fetchActivity: (scroll: ScrollRequest<number>, filter?: ActivityFilter) => void;
    resetActivity: () => void;
    setRefresh: (refresh?: () => void) => void;
    updateFilter: (filter: Partial<ActivityFilter>) => void;
}

export interface ActivityOwnProps {
}

export type ActivityProps = ActivityReduxObject & ActivityDispatchProps & ActivityOwnProps;
