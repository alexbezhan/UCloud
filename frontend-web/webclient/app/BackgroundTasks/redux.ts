import {Speed, Task, TaskUpdate} from "BackgroundTasks/api";
import {Action} from "redux";
import {associateBy, takeLast} from "Utilities/CollectionUtilities";

export interface TaskReduxState {
    tasks?: Record<string ,TaskUpdate>;
}

interface TaskUpdateAction extends Action<"TASK_UPDATE"> {
    update: TaskUpdate;
}

interface TaskLoadAction extends Action<"TASK_LOAD"> {
    page: Page<Task>;
}

export function taskUpdateAction(update: TaskUpdate): TaskUpdateAction {
    return {type: "TASK_UPDATE", update};
}

export function taskLoadAction(page: Page<Task>): TaskLoadAction {
    return {type: "TASK_LOAD", page};
}

type TypeActions = TaskUpdateAction | TaskLoadAction;

export const reducer = (
    state: TaskReduxState = ({}),
    action: TypeActions
): TaskReduxState => {
    switch (action.type) {
        case "TASK_UPDATE": {
            const existingTask: TaskUpdate | undefined = state ? state[action.update.jobId] : undefined;
            if (action.update.complete) {
                const copy = {...state};
                delete copy[action.update.jobId];
                return copy;
            }

            if (!existingTask) {
                const tasks = {...state};
                tasks[action.update.jobId] = {
                    ...action.update,
                    speeds: insertTimestamps(action.update.speeds)
                };
                return tasks;
            } else {
                const currentMessage = existingTask.messageToAppend ? existingTask.messageToAppend : "";
                const messageToAdd = action.update.messageToAppend ? action.update.messageToAppend : "";
                const newMessage = currentMessage + messageToAdd;

                const newStatus = action.update.newStatus ? action.update.newStatus : existingTask.newStatus;
                const newTitle = action.update.newTitle ? action.update.newTitle : existingTask.newTitle;
                const newProgress = action.update.progress ? action.update.progress : existingTask.progress;
                const newSpeed = takeLast((existingTask.speeds || []).concat(action.update.speeds || []), 500);
                const newComplete = action.update.complete ? action.update.complete : existingTask.complete;

                const tasks = {...state};
                tasks[action.update.jobId] = {
                    ...existingTask,
                    messageToAppend: newMessage,
                    progress: newProgress,
                    speeds: insertTimestamps(newSpeed),
                    complete: newComplete,
                    newStatus,
                    newTitle
                };

                return tasks;
            }
        }

        case "TASK_LOAD": {
            const updates: TaskUpdate[] = action.page.items.map(item => {
                return {
                    jobId: item.jobId,
                    speeds: [],
                    messageToAppend: null,
                    progress: null,
                    newTitle: item.title,
                    complete: false,
                    newStatus: item.status
                };
            });

            return associateBy(updates, it => it.jobId);
        }
        default:
            return state;
    }
};

function insertTimestamps(speeds: Speed[]): Speed[] {
    return speeds.map(it => {
        if (it.clientTimestamp) {
            return it;
        } else {
            return {...it, clientTimestamp: Date.now()};
        }
    });
}
