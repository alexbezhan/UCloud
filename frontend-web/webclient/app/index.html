<!DOCTYPE html>
<html lang="en">
<head>
    <meta charset="UTF-8">
    <meta name="viewport" content="width=device-width, initial-scale=1, maximum-scale=1">
    <title>UCloud</title>
</head>
<body>
    <noscript>
        JavaScript is required to run this page.
    </noscript>
    <div id="app"></div>
</body>
</html>
