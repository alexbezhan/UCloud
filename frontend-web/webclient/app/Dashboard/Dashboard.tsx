import * as React from "react";
import {Cloud} from "Authentication/SDUCloudObject"
import {
    favoriteFile,
    getParentPath,
    getFilenameFromPath,
    replaceHomeFolder,
    isDirectory
} from "Utilities/FileUtilities";
import {updatePageTitle, setActivePage} from "Navigation/Redux/StatusActions";
import {
    setAllLoading,
    fetchFavorites,
    fetchRecentAnalyses,
    fetchRecentFiles,
    receiveFavorites
} from "./Redux/DashboardActions";
import {connect} from "react-redux";
import * as moment from "moment";
import {FileIcon} from "UtilityComponents";
import {DashboardProps, DashboardOperations, DashboardStateProps} from ".";
import {Notification, NotificationEntry} from "Notifications";
import {Analysis, AppState} from "Applications";
import {File} from "Files";
import {Dispatch} from "redux";
import {ReduxObject} from "DefaultObjects";
import {Box, Flex, Card, Text, Link, Icon} from "ui-components";
import {EllipsedText} from "ui-components/Text"
import * as Heading from "ui-components/Heading";
import List from "ui-components/List";
import {GridCardGroup} from "ui-components/Grid";
import {fileTablePage} from "Utilities/FileUtilities";
import {notificationRead, readAllNotifications} from "Notifications/Redux/NotificationsActions";
import {History} from "history";
import Spinner from "LoadingIcon/LoadingIcon";
import * as UF from "UtilityFunctions";
import * as Accounting from "Accounting";
import {MainContainer} from "MainContainer/MainContainer";
import {fetchUsage} from "Accounting/Redux/AccountingActions";
import {SidebarPages} from "ui-components/Sidebar";
import {setRefreshFunction} from "Navigation/Redux/HeaderActions";

const DashboardCard: React.FunctionComponent<{title: string, isLoading: boolean}> = ({title, isLoading, children}) => (
    <Card height="auto" width={1} boxShadow="sm" borderWidth={1} borderRadius={6} style={{overflow: "hidden"}}>
        <Flex bg="lightGray" color="darkGray" px={3} py={2} alignItems="center">
            <Heading.h4>{title}</Heading.h4>
        </Flex>
        <Box px={3} py={1}>
            {isLoading && <Spinner size={24} />}
            <Box pb="0.5em" />
            {!isLoading ? children : null}
        </Box>
    </Card>
);

function Dashboard(props: DashboardProps & {history: History}) {

    React.useEffect(() => {
        props.onInit();
        reload(true);
        props.setRefresh(() => reload(true));
        return () => props.setRefresh();
    }, []);

    function reload(loading: boolean) {
        props.setAllLoading(loading);
        props.fetchFavorites();
        props.fetchRecentFiles();
        props.fetchRecentAnalyses();
        props.fetchUsage();
    }

    const onNotificationAction = (notification: Notification) => {
        // FIXME: Not DRY, reused
        switch (notification.type) {
            case "APP_COMPLETE":
                props.history.push(`/applications/results/${notification.meta.jobId}`);
                break;
            case "SHARE_REQUEST":
                props.history.push("/shares");
                break;
        }
    };

    const favoriteOrUnfavorite = (file: File) => {
        favoriteFile(file, Cloud);
        props.receiveFavorites(favoriteFiles.filter(f => f.favorited));
    };

    const {favoriteFiles, recentFiles, recentAnalyses, notifications, favoriteLoading, recentLoading, analysesLoading} = props;
    favoriteFiles.forEach(f => f.favorited = true);
    const main = (
        <>
            <GridCardGroup minmax={290}>
                <DashboardFavoriteFiles
                    files={favoriteFiles}
                    isLoading={favoriteLoading}
                    favorite={file => favoriteOrUnfavorite(file)}
                />

                <DashboardRecentFiles
                    files={recentFiles}
                    isLoading={recentLoading}
                />

                <DashboardAnalyses
                    analyses={recentAnalyses}
                    isLoading={analysesLoading}
                />

                <DashboardNotifications
                    onNotificationAction={onNotificationAction}
                    notifications={notifications}
                    readAll={() => props.readAll()}
                />

                <DashboardCard title={"Storage Used"} isLoading={false}>
                    <Accounting.Usage resource={"storage"} subResource={"bytesUsed"} />
                </DashboardCard>

                <DashboardCard title={"Compute Time Used"} isLoading={false}>
                    <Accounting.Usage resource={"compute"} subResource={"timeUsed"} />
                </DashboardCard>
            </GridCardGroup>
        </>
    );

    return (<MainContainer main={main} />);
}


const DashboardFavoriteFiles = ({files, isLoading, favorite}: {files: File[], isLoading: boolean, favorite: (file: File) => void}) => (
    <DashboardCard title="Favorite Files" isLoading={isLoading}>
        {files.length || isLoading ? null : (<Heading.h6>No favorites found</Heading.h6>)}
        <List>
            {files.map(file => (
                <Flex alignItems="center" key={file.fileId!} pt="0.5em" pb="6.4px">
                    <ListFileContent file={file} link={false} pixelsWide={200} />
                    <Icon ml="auto" size="1em" name="starFilled" color="blue" cursor="pointer" onClick={() => favorite(file)} />
                </Flex>)
            )}
        </List>
    </DashboardCard>
);

const ListFileContent = ({file, link, pixelsWide}: {file: File, link: boolean, pixelsWide: number}) => {
    const iconType = UF.iconFromFilePath(file.path, file.fileType, Cloud.homeFolder);
    return (
        <Flex alignItems="center">
            <FileIcon fileIcon={iconType} link={link} />
            <Link ml="0.5em" to={fileTablePage(isDirectory(file) ? file.path : getParentPath(file.path))}>
                <EllipsedText fontSize={2} width={pixelsWide}>
                    {getFilenameFromPath(replaceHomeFolder(file.path, Cloud.homeFolder))}
                </EllipsedText>
            </Link>
        </Flex>
    );
};

const DashboardRecentFiles = ({files, isLoading}: {files: File[], isLoading: boolean}) => (
    <DashboardCard title="Recently Used Files" isLoading={isLoading}>
        {files.length || isLoading ? null : (<Heading.h6>No recent files found</Heading.h6>)}
        <List>
            {files.map((file, i) => (
                <Flex key={i} alignItems="center" pt="0.5em" pb="0.3em">
                    <ListFileContent file={file} link={file.link} pixelsWide={130} />
                    <Box ml="auto" />
                    <Text fontSize={1} color="grey">{moment(new Date(file.modifiedAt!)).fromNow()}</Text>
                </Flex>
            ))}
        </List>
    </DashboardCard>
);

const DashboardAnalyses = ({analyses, isLoading}: {analyses: Analysis[], isLoading: boolean}) => (
    <DashboardCard title="Recent Jobs" isLoading={isLoading}>
        {isLoading || analyses.length ? null : (<Heading.h6>No results found</Heading.h6>)}
        <List>
            {analyses.map((analysis: Analysis, index: number) =>
                <Flex key={index} alignItems="center" pt="0.5em" pb="8.4px">
                    <Icon name={statusToIconName(analysis.state)}
                        color={statusToColor(analysis.state)}
                        size="1.2em"
                        pr="0.3em"
                    />
                    <Link to={`/applications/results/${analysis.jobId}`}><EllipsedText width={130} fontSize={2}>{analysis.metadata.title}</EllipsedText></Link>
                    <Box ml="auto" />
                    <Text fontSize={1} color="grey">{moment(new Date(analysis.modifiedAt!)).fromNow()}</Text>
                </Flex>
            )}
        </List>
    </DashboardCard>
);

interface DashboardNotificationProps {
    onNotificationAction: (notification: Notification) => void
    notifications: Notification[]
    readAll: () => void
}

const DashboardNotifications = ({notifications, readAll, onNotificationAction}: DashboardNotificationProps) => (
    <Card height="auto" width={1} boxShadow="sm" borderWidth={1} borderRadius={6} style={{overflow: "hidden"}}>
        <Flex bg="lightGray" color="darkGray" px={3} py={2}>
            <Heading.h4>Recent Notifications</Heading.h4>
            <Box ml="auto" />
            <Icon name="checkDouble" cursor="pointer" color="iconColor" color2="iconColor2" title="Mark all as read" onClick={readAll} />
        </Flex>
        {notifications.length === 0 ? <Heading.h6 pl={"16px"} pt="10px">No notifications</Heading.h6> : null}
        <List>
            {notifications.slice(0, 7).map((n, i) =>
                <Flex key={i}>
                    <NotificationEntry notification={n} onAction={onNotificationAction} />
                </Flex>
            )}
        </List>
    </Card>
);

const statusToIconName = (status: AppState) => {
    switch (status) {
        case AppState.SUCCESS:
            return "check";
        case AppState.FAILURE:
            return "close";
        case AppState.SCHEDULED:
            return "calendar";
        case AppState.RUNNING:
            return "chrono";
        case AppState.VALIDATED:
            return "checkDouble";
        default:
            return "ellipsis";
    }
};

const statusToColor = (status: AppState) => status === AppState.FAILURE ? "red" : "green";

const mapDispatchToProps = (dispatch: Dispatch): DashboardOperations => ({
    onInit: () => {
        dispatch(updatePageTitle("Dashboard"));
        dispatch(setActivePage(SidebarPages.None));
    },
    setAllLoading: loading => dispatch(setAllLoading(loading)),
    fetchFavorites: async () => dispatch(await fetchFavorites()),
    fetchRecentFiles: async () => dispatch(await fetchRecentFiles()),
    fetchRecentAnalyses: async () => dispatch(await fetchRecentAnalyses()),
    fetchUsage: async () => {
        dispatch(await fetchUsage("storage", "bytesUsed"));
        dispatch(await fetchUsage("compute", "timeUsed"));
    },
    notificationRead: async id => dispatch(await notificationRead(id)),
    readAll: async () => dispatch(await readAllNotifications()),
    // FIXME: Make action instead (favoriteFile)
    receiveFavorites: files => dispatch(receiveFavorites(files)),
    setRefresh: refresh => dispatch(setRefreshFunction(refresh))
});

const mapStateToProps = (state: ReduxObject): DashboardStateProps => ({
    ...state.dashboard,
    notifications: state.notifications.items,
    favoriteFilesLength: state.dashboard.favoriteFiles.length // Hack to ensure re-rendering
});

export default connect(mapStateToProps, mapDispatchToProps)(Dashboard);
