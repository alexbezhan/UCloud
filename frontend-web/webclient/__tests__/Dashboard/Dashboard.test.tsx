import {createMemoryHistory} from "history";
import * as React from "react";
import {Provider} from "react-redux";
import {MemoryRouter} from "react-router";
import {create} from "react-test-renderer";
import {Store} from "redux";
import {ThemeProvider} from "styled-components";
import Dashboard from "../../app/Dashboard/Dashboard";
import {ReduxObject} from "../../app/DefaultObjects";
import theme from "../../app/ui-components/theme";
import {store} from "../../app/Utilities/ReduxUtilities";

const WrappedDashboard: React.FunctionComponent<{store: Store<ReduxObject>}> = props => {
    return (
        <Provider store={props.store}>
            <ThemeProvider theme={theme}>
                <MemoryRouter>
                    <Dashboard history={createMemoryHistory()} />
                </MemoryRouter>
            </ThemeProvider>
        </Provider>
    );
};

describe("Dashboard", () => {
    test("Dashboard mount", () => {
        expect(create(<WrappedDashboard store={store} />)).toMatchSnapshot();
    });
});
