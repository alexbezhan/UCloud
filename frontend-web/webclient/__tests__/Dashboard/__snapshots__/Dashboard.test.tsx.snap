// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dashboard Dashboard mount 1`] = `
.c18 {
  width: 32px;
  height: 32px;
  margin: 20px auto;
  -webkit-animation-name: jkjquP;
  animation-name: jkjquP;
}

.c18 > svg path:nth-child(1) {
  -webkit-animation: jkjquP 0.72s linear infinite;
  animation: jkjquP 0.72s linear infinite;
  -webkit-animation-delay: -0.04s;
  animation-delay: -0.04s;
}

.c18 > svg path:nth-child(2) {
  -webkit-animation: jkjquP 0.72s linear infinite;
  animation: jkjquP 0.72s linear infinite;
  -webkit-animation-delay: -0.08s;
  animation-delay: -0.08s;
}

.c18 > svg path:nth-child(3) {
  -webkit-animation: jkjquP 0.72s linear infinite;
  animation: jkjquP 0.72s linear infinite;
  -webkit-animation-delay: -0.12s;
  animation-delay: -0.12s;
}

.c18 > svg path:nth-child(4) {
  -webkit-animation: jkjquP 0.72s linear infinite;
  animation: jkjquP 0.72s linear infinite;
  -webkit-animation-delay: -0.16s;
  animation-delay: -0.16s;
}

.c18 > svg path:nth-child(5) {
  -webkit-animation: jkjquP 0.72s linear infinite;
  animation: jkjquP 0.72s linear infinite;
  -webkit-animation-delay: -0.2s;
  animation-delay: -0.2s;
}

.c18 > svg path:nth-child(6) {
  -webkit-animation: jkjquP 0.72s linear infinite;
  animation: jkjquP 0.72s linear infinite;
  -webkit-animation-delay: -0.24s;
  animation-delay: -0.24s;
}

.c18 > svg path:nth-child(7) {
  -webkit-animation: jkjquP 0.72s linear infinite;
  animation: jkjquP 0.72s linear infinite;
  -webkit-animation-delay: -0.28s;
  animation-delay: -0.28s;
}

.c18 > svg path:nth-child(8) {
  -webkit-animation: jkjquP 0.72s linear infinite;
  animation: jkjquP 0.72s linear infinite;
  -webkit-animation-delay: -0.32s;
  animation-delay: -0.32s;
}

.c18 > svg path:nth-child(9) {
  -webkit-animation: jkjquP 0.72s linear infinite;
  animation: jkjquP 0.72s linear infinite;
  -webkit-animation-delay: -0.36s;
  animation-delay: -0.36s;
}

.c18 > svg path:nth-child(10) {
  -webkit-animation: jkjquP 0.72s linear infinite;
  animation: jkjquP 0.72s linear infinite;
  -webkit-animation-delay: -0.4s;
  animation-delay: -0.4s;
}

.c18 > svg path:nth-child(11) {
  -webkit-animation: jkjquP 0.72s linear infinite;
  animation: jkjquP 0.72s linear infinite;
  -webkit-animation-delay: -0.44s;
  animation-delay: -0.44s;
}

.c18 > svg path:nth-child(12) {
  -webkit-animation: jkjquP 0.72s linear infinite;
  animation: jkjquP 0.72s linear infinite;
  -webkit-animation-delay: -0.48s;
  animation-delay: -0.48s;
}

.c18 > svg path:nth-child(13) {
  -webkit-animation: jkjquP 0.72s linear infinite;
  animation: jkjquP 0.72s linear infinite;
  -webkit-animation-delay: -0.52s;
  animation-delay: -0.52s;
}

.c18 > svg path:nth-child(14) {
  -webkit-animation: jkjquP 0.72s linear infinite;
  animation: jkjquP 0.72s linear infinite;
  -webkit-animation-delay: -0.56s;
  animation-delay: -0.56s;
}

.c18 > svg path:nth-child(15) {
  -webkit-animation: jkjquP 0.72s linear infinite;
  animation: jkjquP 0.72s linear infinite;
  -webkit-animation-delay: -0.6s;
  animation-delay: -0.6s;
}

.c18 > svg path:nth-child(16) {
  -webkit-animation: jkjquP 0.72s linear infinite;
  animation: jkjquP 0.72s linear infinite;
  -webkit-animation-delay: -0.64s;
  animation-delay: -0.64s;
}

.c18 > svg path:nth-child(17) {
  -webkit-animation: jkjquP 0.72s linear infinite;
  animation: jkjquP 0.72s linear infinite;
  -webkit-animation-delay: -0.68s;
  animation-delay: -0.68s;
}

.c18 > svg path:nth-child(18) {
  -webkit-animation: jkjquP 0.72s linear infinite;
  animation: jkjquP 0.72s linear infinite;
  -webkit-animation-delay: -0.72s;
  animation-delay: -0.72s;
}

.c0 {
  margin-left: 190px;
  padding-top: 48px;
  padding-bottom: 14px;
  padding-left: 14px;
  padding-right: 0;
  background-color: #fff;
}

.c1 {
  padding-top: 14px;
  padding-right: 14px;
}

.c7 {
  padding-left: 16px;
  padding-right: 16px;
  padding-top: 4px;
  padding-bottom: 4px;
}

.c8 {
  padding-bottom: 0.5em;
}

.c9 {
  text-align: center;
}

.c15 {
  margin-left: auto;
}

.c19 {
  padding-bottom: 12px;
}

.c12 {
  font-smoothing: antialiased;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  text-align: center;
  -webkit-text-decoration: none;
  text-decoration: none;
  font-family: inherit;
  font-weight: 700;
  line-height: 1.5;
  cursor: pointer;
  border-radius: 5px;
  background-color: var(--blue,#f00);
  color: var(--white,#f00);
  border-width: 0;
  border-style: solid;
  -webkit-transition: cubic-bezier(0.5,0,0.25,1) 60ms;
  transition: cubic-bezier(0.5,0,0.25,1) 60ms;
  font-size: 14px;
  padding: 9.5px 18px;
}

.c12:disabled {
  opacity: 0.25;
}

.c12:focus {
  outline: none;
}

.c12:hover {
  -webkit-transform: scale(1.03);
  -ms-transform: scale(1.03);
  transform: scale(1.03);
}

.c16 {
  -webkit-flex: none;
  -ms-flex: none;
  flex: none;
  vertical-align: middle;
  cursor: pointer;
  color: #53657d;
}

.c4 {
  width: 100%;
  height: auto;
  overflow: hidden;
  height: auto;
  box-shadow: 0px 3px 3px -2px rgba(0,0,0,0.2),0px 3px 4px 0px rgba(0,0,0,.14),0px 1px 8px 0px rgba(0,0,0,.12);
  border: 1px solid #c9d3df;
  border-radius: 6px;
}

.c5 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  padding-left: 16px;
  padding-right: 16px;
  padding-top: 8px;
  padding-bottom: 8px;
  background-color: #f5f7f9;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c14 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  padding-left: 16px;
  padding-right: 16px;
  padding-top: 8px;
  padding-bottom: 8px;
  background-color: #f5f7f9;
}

.c3 {
  display: grid;
  margin-top: 2px;
  width: 100%;
  grid-gap: 10px;
  grid-template-columns: repeat(auto-fill,minmax(315px,1fr) );
}

.c6 {
  cursor: inherit;
  font-size: 20px;
  margin: 0;
  font-weight: 400;
}

.c10 {
  cursor: inherit;
  font-size: 16px;
  margin-top: 8px;
  margin-bottom: 8px;
}

.c17 {
  cursor: inherit;
  font-size: 10px;
  margin: 0;
  padding-left: 16px;
  padding-top: 10px;
  text-transform: uppercase;
  font-weight: 700;
}

.c11 {
  cursor: pointer;
  -webkit-text-decoration: none;
  text-decoration: none;
  color: #1e252e;
}

.c11:hover {
  color: #006aff;
}

.c13 {
  font-size: large;
}

.c13 > * {
  border-bottom: 1px solid lightGrey;
}

.c13 > *:last-child {
  border-bottom: 0px;
}

@media screen and (min-width:768px) and (max-width:1023px) {
  .c2 {
    display: none;
  }
}

@media screen and (min-width:1024px) and (max-width:1279px) {
  .c2 {
    display: none;
  }
}

@media screen and (min-width:1280px) {
  .c2 {
    display: none;
  }
}

<div
  className="c0"
>
  <div
    className="c1"
  >
    <div
      className="c2"
    />
    <div
      className="c3"
      width="100%"
    >
      <div
        className="c4"
        height="auto"
        overflow="hidden"
        width={1}
      >
        <div
          className="c5"
        >
          <h4
            className="c6"
            cursor="inherit"
            fontSize={3}
          >
            Favorite Files
          </h4>
        </div>
        <div
          className="c7"
        >
          <div
            className="c8"
          />
          <div
            className="c9"
          >
            <div
              className="c10"
              cursor="inherit"
              fontSize="16px"
            >
              Your favorite files will appear here
            </div>
            <a
              className="c11"
              color="text"
              hoverColor="textHighlight"
              href="/files?path=%2Fhome%2Ftest%40test.dk"
              onClick={[Function]}
            >
              <button
                className="c12"
                color="blue"
              >
                Explore files
              </button>
            </a>
          </div>
          <div
            className="c13"
            fontSize="large"
          />
        </div>
      </div>
      <div
        className="c4"
        height="auto"
        overflow="hidden"
        width={1}
      >
        <div
          className="c5"
        >
          <h4
            className="c6"
            cursor="inherit"
            fontSize={3}
          >
            Recent Jobs
          </h4>
        </div>
        <div
          className="c7"
        >
          <div
            className="c8"
          />
          <div
            className="c9"
          >
            <div
              className="c10"
              cursor="inherit"
              fontSize="16px"
            >
              No recent jobs
            </div>
            <a
              className="c11"
              color="text"
              hoverColor="textHighlight"
              href="/applications/overview"
              onClick={[Function]}
            >
              <button
                className="c12"
                color="blue"
              >
                Explore apps
              </button>
            </a>
          </div>
          <div
            className="c13"
            fontSize="large"
          />
        </div>
      </div>
      <div
        className="c4"
        height="auto"
        overflow="hidden"
        width={1}
      >
        <div
          className="c14"
        >
          <h4
            className="c6"
            cursor="inherit"
            fontSize={3}
          >
            Recent Notifications
          </h4>
          <div
            className="c15"
          />
          <svg
            className="c16"
            clipRule="evenodd"
            color="iconColor"
            color2=""
            cursor="pointer"
            fill="currentcolor"
            fillRule="evenodd"
            height={24}
            onClick={[Function]}
            title="Mark all as read"
            viewBox="0 0 25 19"
            width={24}
          >
            <path
              d="M7.999 15l-6-6.019-2 2.02 8 8 16-17-2.001-2-14 15z"
              fillRule="nonzero"
            />
            <path
              d="M13 16.962L14.999 19 24 9.96l-2-1.999-9 9z"
              fill={null}
              fillRule="nonzero"
            />
          </svg>
        </div>
        <h6
          className="c17"
          cursor="inherit"
          fontSize={0}
        >
          No notifications
        </h6>
        <div
          className="c13"
          fontSize="large"
        />
      </div>
      <div
        className="c4"
        height="auto"
        overflow="hidden"
        width={1}
      >
        <div
          className="c5"
        >
          <h4
            className="c6"
            cursor="inherit"
            fontSize={3}
          >
            Resources
          </h4>
        </div>
        <div
          className="c7"
        >
          <div
            className="c8"
          />
          <div
            className="c9"
          >
            <div
              className="c18"
              height={32}
              width={32}
            >
              <svg
                height={32}
                id="svg"
                version="1.1"
                viewBox="0 0 220 220"
                width={32}
                xmlns="http://www.w3.org/2000/svg"
                xmlnsXlink="http://www.w3.org/1999/xlink"
              >
                <path
                  d="M80.296,0.12l14.673,54.761l40.088,-40.088l-54.761,-14.673Z"
                />
                <path
                  d="M149.73,69.554l-14.673,-54.761l-40.088,40.088l54.761,14.673Z"
                />
                <path
                  d="M189.818,29.466l-54.761,-14.673l14.673,54.761l40.088,-40.088Z"
                />
                <path
                  d="M204.491,84.228l-14.673,-54.762l-40.088,40.088l54.761,14.674Z"
                />
                <path
                  d="M164.403,124.316l40.088,-40.088l-54.761,-14.674l14.673,54.762Z"
                />
                <path
                  d="M219.165,138.989l-14.674,-54.761l-40.088,40.088l54.762,14.673Z"
                />
                <path
                  d="M179.077,179.077l-14.674,-54.761l54.762,14.673l-40.088,40.088Z"
                />
                <path
                  d="M124.316,164.403l54.761,14.674l-14.674,-54.761l-40.087,40.087Z"
                />
                <path
                  d="M138.989,219.165l40.088,-40.088l-54.761,-14.674l14.673,54.762Z"
                />
                <path
                  d="M84.228,204.491l54.761,14.674l-14.673,-54.762l-40.088,40.088Z"
                />
                <path
                  d="M69.554,149.73l14.674,54.761l40.088,-40.088l-54.762,-14.673Z"
                />
                <path
                  d="M29.466,189.818l54.762,14.673l-14.674,-54.761l-40.088,40.088Z"
                />
                <path
                  d="M14.793,135.057l14.673,54.761l40.088,-40.088l-54.761,-14.673Z"
                />
                <path
                  d="M54.881,94.969l-40.088,40.088l54.761,14.673l-14.673,-54.761Z"
                />
                <path
                  d="M0.12,80.296l14.673,54.761l40.088,-40.088l-54.761,-14.673Z"
                />
                <path
                  d="M40.208,40.208l-40.088,40.088l54.761,14.673l-14.673,-54.761Z"
                />
                <path
                  d="M94.969,54.881l-54.761,-14.673l14.673,54.761l40.088,-40.088Z"
                />
                <path
                  d="M80.296,0.12l-40.088,40.088l54.761,14.673l-14.673,-54.761Z"
                />
              </svg>
            </div>
          </div>
          <div
            className="c19"
          />
          <div
            className="c9"
          >
            <div
              className="c18"
              height={32}
              width={32}
            >
              <svg
                height={32}
                id="svg"
                version="1.1"
                viewBox="0 0 220 220"
                width={32}
                xmlns="http://www.w3.org/2000/svg"
                xmlnsXlink="http://www.w3.org/1999/xlink"
              >
                <path
                  d="M80.296,0.12l14.673,54.761l40.088,-40.088l-54.761,-14.673Z"
                />
                <path
                  d="M149.73,69.554l-14.673,-54.761l-40.088,40.088l54.761,14.673Z"
                />
                <path
                  d="M189.818,29.466l-54.761,-14.673l14.673,54.761l40.088,-40.088Z"
                />
                <path
                  d="M204.491,84.228l-14.673,-54.762l-40.088,40.088l54.761,14.674Z"
                />
                <path
                  d="M164.403,124.316l40.088,-40.088l-54.761,-14.674l14.673,54.762Z"
                />
                <path
                  d="M219.165,138.989l-14.674,-54.761l-40.088,40.088l54.762,14.673Z"
                />
                <path
                  d="M179.077,179.077l-14.674,-54.761l54.762,14.673l-40.088,40.088Z"
                />
                <path
                  d="M124.316,164.403l54.761,14.674l-14.674,-54.761l-40.087,40.087Z"
                />
                <path
                  d="M138.989,219.165l40.088,-40.088l-54.761,-14.674l14.673,54.762Z"
                />
                <path
                  d="M84.228,204.491l54.761,14.674l-14.673,-54.762l-40.088,40.088Z"
                />
                <path
                  d="M69.554,149.73l14.674,54.761l40.088,-40.088l-54.762,-14.673Z"
                />
                <path
                  d="M29.466,189.818l54.762,14.673l-14.674,-54.761l-40.088,40.088Z"
                />
                <path
                  d="M14.793,135.057l14.673,54.761l40.088,-40.088l-54.761,-14.673Z"
                />
                <path
                  d="M54.881,94.969l-40.088,40.088l54.761,14.673l-14.673,-54.761Z"
                />
                <path
                  d="M0.12,80.296l14.673,54.761l40.088,-40.088l-54.761,-14.673Z"
                />
                <path
                  d="M40.208,40.208l-40.088,40.088l54.761,14.673l-14.673,-54.761Z"
                />
                <path
                  d="M94.969,54.881l-54.761,-14.673l14.673,54.761l40.088,-40.088Z"
                />
                <path
                  d="M80.296,0.12l-40.088,40.088l54.761,14.673l-14.673,-54.761Z"
                />
              </svg>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
